### PR TITLE
Updating Modal to a function component within react-next

### LINKED ISSUE #14297: Modal: strict mode compliance
Modal has a moderately complex `componentWillReceiveProps` implementation which does the following:
- starting open/close animation
- setting up dragging
- layout measurement

Given time constraints, fixing this in place rather than converting to a function component might be a better idea.

### PR DESCRIPTION
**Pull request by czearing**
#### Pull request checklist
- [X] Include a change request file using `$ yarn change`
- [X] Fixes #14297

#### Description of changes

1. Updating `Modal` to a function component within react-next.
2. Adding perf and vr tests for `Modal` to app package.

#### Focus areas to test
`Modal`

### CODE REVIEW

File: packages/react-next/src/components/Modal/Modal.base.tsx
Comment (xugao) [inline packages/react-next/src/components/Modal/Modal.base.tsx:313]:
```suggestion
      setModalMenuClose();
```

File: packages/react-next/src/components/Modal/Modal.base.tsx
Comment (xugao) [inline packages/react-next/src/components/Modal/Modal.base.tsx:144]: we should `useCallback` on this

**Review by xugao — COMMENTED** [review]
I know we didn't plan to convert `Dialog` to FC. But i think we should do it (after this PR).
I looked at the dialog code it's going to be a straight-forward conversion :) Mainly I also want to remove `IWithResponsiveState` prop from `Dialog` to keep it more consistent with `Modal`.

File: packages/react-next/src/components/Modal/Modal.base.tsx
Comment (czearing) [reply packages/react-next/src/components/Modal/Modal.base.tsx:110]: It's passed into getClassNames. When Modal is opened for the first time and the topOffsetFixed prop is true it will set root's alignItems to 'flex-start' and set main's top to modalRectangleTop.

**Review by xugao — APPROVED** [review]

File: packages/examples/src/react-next/__snapshots__/Keytips.CommandBar.Example.tsx.shot
Comment (ecraig12345) [inline packages/examples/src/react-next/__snapshots__/Keytips.CommandBar.Example.tsx.shot:675]: Did this PR somehow accidentally change Modal so it's by default? That would be my guess from looking at this snapshot.

File: packages/react-next/src/components/Modal/Modal.base.tsx
Comment (ecraig12345) [inline packages/react-next/src/components/Modal/Modal.base.tsx:106]: In this case it's probably better to just do `useState`

File: packages/react-next/src/components/Modal/Modal.base.tsx
Comment (ecraig12345) [inline packages/react-next/src/components/Modal/Modal.base.tsx:108]: Arguably this should also just use `useState`. Toggling can be handled without direct access to the current value by doing `setIsModalMenuOpen(prevValue => !prevValue)`. Though since you're using the toggle callback frequently I could go either way on this one.

File: packages/react-next/src/components/Modal/Modal.base.tsx
Comment (ecraig12345) [inline packages/react-next/src/components/Modal/Modal.base.tsx:110]: This should definitely just do `useState`.

File: packages/react-next/src/components/Modal/Modal.base.tsx
Comment (ecraig12345) [inline packages/react-next/src/components/Modal/Modal.base.tsx:112]: Same here about useState

File: packages/react-next/src/components/Modal/Modal.base.tsx
Comment (ecraig12345) [inline packages/react-next/src/components/Modal/Modal.base.tsx:114]: And here

**Review by ecraig12345 — COMMENTED** [review]
There are several cases here where I think it's more appropriate to do `useState` instead of `useBoolean`. `useBoolean` is mainly meant for when you want to pass a particular setter directly as a callback. Native `useState` will be a bit cheaper and also has the benefit that you don't need to pass it as a dependency everywhere to satisfy the eslint rule (it has a hardcoded exception since it knows the setter from useState will never change).

**Comment by ecraig12345** [comment]
Also a general FYI, if a PR fully fixes a particular issue you should put `Fixes #nnnn` so the issue will be automatically closed.

File: packages/examples/src/react-next/__snapshots__/Keytips.CommandBar.Example.tsx.shot
Comment (czearing) [reply packages/examples/src/react-next/__snapshots__/Keytips.CommandBar.Example.tsx.shot]: Thanks for mentioning this! There turned out to be a bug in `Modal's` dismissal.

File: packages/react-next/src/components/Modal/Modal.base.tsx
Comment (ecraig12345) [inline packages/react-next/src/components/Modal/Modal.base.tsx:350]: This may not be a new issue but these items ought to be memoized (and you can move the callback definitions inside the memoization, so they don't need to be memoized separately)

**Review by ecraig12345 — CHANGES_REQUESTED** [review]
I started making changes to fix a couple things I noticed while taking a last pass over this, but then as I got into it things turned out to be more complicated than I thought (like my attempted fixes kept causing or exposing more bugs). I'll add a detailed review later but basically let's wait on this one until after we do the OUFR/react-next merge.

**Pull request closed without merging**
